Add Doc Samples For CheckboxListTile, RadioListTile and SwitchListTile

shihaohong opened the pull request:
## Description

Adds samples to ListTile variants so that developers have some structure to go off of if they decide to create their own custom [Checkbox/Radio/Switch]Label widgets. The goal is to emphasize that ListTiles aren't a one-size-fits-all solution and that it's possible to create a custom widget that better suits developers' needs.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/30781

## Tests

No tests since these are code samples in the API documentation.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

shihaohong committed:
Moved Radio documentation line to be above sample

shihaohong committed:
Added LabeledRadio sample

shihaohong committed:
Improve sentence in doc

shihaohong committed:
Add missing comma

shihaohong committed:
Add LabeledCheckbox sample

shihaohong committed:
Add LabeledSwitch sample

HansMuller commented on packages/flutter/lib/src/material/checkbox_list_tile.dart:
```diff
@@ -57,6 +57,78 @@ import 'theme_data.dart';
 ///
+/// ## CheckboxListTile isn't exactly what I want
+///
+/// If the way ListTile pads and positions its elements isn't quite what you're
```
ListTile => CheckBoxListTile?

HansMuller commented on packages/flutter/lib/src/material/checkbox_list_tile.dart:
```diff
@@ -57,6 +57,78 @@ import 'theme_data.dart';
+/// ## CheckboxListTile isn't exactly what I want
+///
+/// If the way ListTile pads and positions its elements isn't quite what you're
+/// looking for, you can create custom labeled radio widgets by combining it
```
"create custom labeled radio widgets" aren't we talking about check boxes here?

combining it => combining [Checkbox]

HansMuller commented on packages/flutter/lib/src/material/checkbox_list_tile.dart:
```diff
@@ -57,6 +57,78 @@ import 'theme_data.dart';
+///       onTap: () {
+///         onChanged(!value);
+///       },
+///       child: Container(
```
Might as well use Padding instead of Container here

HansMuller commented on packages/flutter/lib/src/material/radio_list_tile.dart:
```diff
@@ -72,6 +72,81 @@ import 'theme_data.dart';
 ///
+/// ## RadioListTile isn't exactly what I want
+///
+/// If the way ListTile pads and positions its elements isn't quite what you're
```
ListTile => RadioListTile?

HansMuller commented on packages/flutter/lib/src/material/radio_list_tile.dart:
```diff
@@ -72,6 +72,81 @@ import 'theme_data.dart';
+/// ## RadioListTile isn't exactly what I want
+///
+/// If the way ListTile pads and positions its elements isn't quite what you're
+/// looking for, you can create custom labeled radio widgets by combining it
```
combining it => combining [Radio]

HansMuller commented on packages/flutter/lib/src/material/radio_list_tile.dart:
```diff
@@ -72,6 +72,81 @@ import 'theme_data.dart';
+///         if (value == false)
+///           onChanged(true);
+///       },
+///       child: Container(
```
Container => Padding

HansMuller commented on packages/flutter/lib/src/material/switch_list_tile.dart:
```diff
@@ -67,6 +67,80 @@ enum _SwitchListTileType { material, adaptive }
 ///
+/// ## SwitchListTile isn't exactly what I want
+///
+/// If the way ListTile pads and positions its elements isn't quite what you're
```
Same comments as before

shihaohong committed:
Update doc wording

shihaohong committed:
Switch Container to Padding in examples

HansMuller reviewed (dismissed):
LGTM

shihaohong committed:
Update SwitchListTile to include tappable RichText

shihaohong committed:
Refactor LabeledRadio to be LabeledLinkedRadio

shihaohong committed:
Remove InkWell from SwitchListTile example

shihaohong committed:
Update to Checkbox RichText Example

shihaohong committed:
Fix incorrect usage of checkbox widgets in switch sample

shihaohong committed:
Update helper widget references

shihaohong commented:
Made some minor updates to incorporate @chunhtai's example with `RichText`. Please review, thanks!

shihaohong commented:
The analyzer test is failing because there's a bug in the API doc sample generator that does not properly insert `import 'packages:flutter/gestures.dart` for the app sample. I filed an issue to track it: https://github.com/flutter/flutter/issues/32725

chunhtai commented on packages/flutter/lib/src/material/checkbox_list_tile.dart:
````diff
@@ -57,6 +57,88 @@ import 'theme_data.dart';
+/// import 'package:flutter/gestures.dart';
+/// ```
+/// ```dart preamble
+/// class LabeledLinkedCheckbox extends StatelessWidget {
````
I didn't think twice when i made up this name, LinkedLabelCheckbox may be a better name.

shihaohong replied:
I'll change it, I think it's better if the readers are less confused by the naming :smile:

chunhtai commented:
LGTM

shihaohong committed:
Fix sample code import block

shihaohong committed:
Improve code sample class names

shihaohong committed:
Added explanation to Semantics samples

shihaohong commented:
It turns out that it wasn't an issue with the snippet generator tool, but an error I made with the snippet I was generating. I made the fix and also added an explanation to the Semantic examples. It's now ready for re-review.

shihaohong committed:
Fix accidental paragraph deletion

shihaohong committed:
Added second example to RadioListTile

shihaohong committed:
Added second example to CheckboxListTile

shihaohong committed:
Fix location of tool sample starting point

shihaohong committed:
Added second example to SwitchListTile

shihaohong committed:
Fix CustomLabeledRadio Example

shihaohong committed:
Fix LabeledRadio example to use groupValue correctly on InkWell tap

shihaohong commented:
Decided offline with @HansMuller to create two separate samples: one to address the Semantics example and one to address the [Switch/Radio/Checkbox] + Label sample issue. That way, there's less confusion from trying to combine the two takeaways into one sample.

HansMuller commented on packages/flutter/lib/src/material/checkbox_list_tile.dart:
```diff
@@ -57,6 +57,165 @@ import 'theme_data.dart';
+///               ),
+///               recognizer: TapGestureRecognizer()
+///                 ..onTap = () {
+///                 print('Link has been tapped.');
```
Maybe 'Label has been tapped'

HansMuller commented on packages/flutter/lib/src/material/checkbox_list_tile.dart:
```diff
@@ -57,6 +57,165 @@ import 'theme_data.dart';
+/// ## Semantics in CheckboxListTile
+///
+/// Since CheckboxListTile is wrapped with a [MergeSemantics] widget, it wants
+/// to merge its children semantics nodes into one node in the semantics tree.
```
its descendant [Semantics] nodes

here and elsewhere

HansMuller commented on packages/flutter/lib/src/material/checkbox_list_tile.dart:
```diff
@@ -57,6 +57,165 @@ import 'theme_data.dart';
+/// Therefore, CheckboxListTile would not work with [RichText], which forces
+/// itself to be a standalone semantics node.
+///
+/// If you're interested in learning more about this interaction, see the
```
I don't see anything about Semantics in https://docs.flutter.io/flutter/painting/TextSpan/recognizer.html

Maybe leave this paragraph out?

chunhtai replied:
This line cause it to become independent semantics node https://github.com/flutter/flutter/blob/4e1bfca8470ff088ecb68aaa72086f290b316fd1/packages/flutter/lib/src/rendering/paragraph.dart#L504

shihaohong replied:
Yeah, I knew it was in `RenderParagraph`, but I just didn't know how to correlate the ideas well, since `RenderParagraph` is utilized by `RichText` under the hood and I didn't want the reader to have to dig too deep to get the idea. I'm leaning on just leaving it out and giving a high level explanation.

HansMuller commented on packages/flutter/lib/src/material/checkbox_list_tile.dart:
```diff
@@ -57,6 +57,165 @@ import 'theme_data.dart';
+/// {@tool snippet --template=stateful_widget_scaffold}
+///
+/// Here is an example of a LinkedLabelCheckbox widget that allows a
+/// [RichText] widget to be nested in a checkbox tile widget.
```
an interactive [RichText] widget...

HansMuller reviewed (commented):
This is turning into a nice big PR!

Feedback is just a few minor wording suggestions. Most of them should apply to all of the samples.

shihaohong committed:
Update description of linked label widgets

shihaohong committed:
Update print string in linked label examples

shihaohong committed:
Added reference to Semantics docs

shihaohong committed:
Update wording of RichText reference in sample

shihaohong committed:
Update semantics paragraph explanation

shihaohong commented on packages/flutter/lib/src/material/checkbox_list_tile.dart:
```diff
@@ -57,6 +57,96 @@ import 'theme_data.dart';
+/// Here is an example of a custom LinkedLabelCheckbox widget that allows a
+/// [RichText] widget to be nested in a checkbox tile widget.
+///
+/// Since CheckboxListTile is wrapped with a [MergeSemantics] widget, it wants
```
I updated it to be the following:
```
/// Since CheckboxListTile wraps its children with a [MergeSemantics] widget, it
/// wants to merge its descendant [Semantics] nodes to appear as just one
/// interactive node in the semantics tree.
///
/// Thus, CheckboxListTile will not work with children widgets that need to be \
/// a standalone interactive element. One example would be the [RichText]
/// widget, which has an embedded gesture recognizer configured to have
/// its own [Semantics] node. Thus, you may need to create a custom widget to
/// accommodate your needs.
```

shihaohong committed:
Reverse order of switch/checkbox with label in samples

shihaohong committed:
increased padding, made const padding

shihaohong committed:
code cleanup

shihaohong committed:
Improve semantics section description

shihaohong committed:
Improve simple Switch/Radio/CheckboxListTile samples

shihaohong committed:
Added assets to samples

shihaohong commented:
Did the following:
1) Added in image examples
2) Made the first sample on each page an entire app to help with copy/pasting to test
3) Fixed up semantics paragraphs to have a cleaner description

HansMuller commented on packages/flutter/lib/src/material/checkbox_list_tile.dart:
```diff
@@ -37,40 +37,57 @@ import 'theme_data.dart';
+/// nodes into one node in the semantics tree. Therefore, CheckboxListTile will
+/// throw an error if any of its children requires its own [Semantics] node.
+///
+/// For example, you cannot nest a [RichText] widget as a descendant of
```
NICE

HansMuller reviewed (approved):
LGTM
